src/main/java/net/rptools/maptool/model/Token.java

```diff
@@ -916,14 +922,30 @@ public String getImageTableName() {
     return imageTableName;
   }
 
-  public void addLightSource(LightSource source) {
-    lightSourceList.add(new AttachedLightSource(source));
+  public @Nonnull Collection<LightSource> getUniqueLightSources() {
+    return uniqueLightSources.values();
+  }
+
+  public @Nullable LightSource getUniqueLightSource(GUID lightSourceId) {
+    return uniqueLightSources.getOrDefault(lightSourceId, null);
+  }
+
+  public void addUniqueLightSource(LightSource source) {
+    uniqueLightSources.put(source.getId(), source);
+  }
+
+  public void removeUniqueLightSource(GUID lightSourceId) {
+    uniqueLightSources.remove(lightSourceId);
+  }
+
+  public void addLightSource(GUID lightSourceId) {
+    lightSourceList.add(new AttachedLightSource(lightSourceId));
   }
 
   public void removeLightSourceType(LightSource.Type lightType) {
     for (ListIterator<AttachedLightSource> i = lightSourceList.listIterator(); i.hasNext(); ) {
```

Comment:
We might es well refactor those loops to use foreach syntax IMO.

Reply:
foreach is more unwieldy here since it doesn't support removal. We would have to build up a list of things to remove, then do the removal separately.